Use `teal.data::get_code` instead of `teal.data:::get_code_dependency`

m7pr opened the pull request:
After introduction of `...` in `teal.data::get_code()` during https://github.com/insightsengineering/teal.data/pull/290,
I reckon there is no point using non-exported `teal.data:::get_code_dependency` 
and we can just use exported `teal.data::get_code`.

Merge only when this is discussed and updated version of `teal.data` lands on CRAN https://github.com/insightsengineering/teal.data/issues/291

m7pr commented:
Review only after this is merged https://github.com/insightsengineering/teal.data/pull/290

averissimo commented:
Getting this in before {teal} is on CRAN would be nice to have 1 less `getFromNamespace()`

averissimo reviewed (changes requested):
Blocking as requested by @m7pr 

Getting this in before {teal} is on CRAN would be nice to have 1 less `getFromNamespace()` in the package

m7pr commented:
https://github.com/insightsengineering/teal.data/pull/290 is merged and `{teal.data}` 0.5.0 is released on https://github.com/insightsengineering/teal.data/pull/297/files

I guess we can merge this one. Restarting GHA builds

m7pr commented:
3 tests are failing

m7pr commented:
I think the quickest fix would be to create a `character` method for `teal.data::get_code` that would work directly on `character` code. So you could do as follows

- `teal.data::get_code(teal_data())  # using method for teal_data object`
- `teal.data::get_code(teal_data()@code) # using method for character`

chlebowa commented:
> method for `teal.data::get_code` that would work directly on `character` code

Bad idea.

Let's discuss on a call.

m7pr commented:
Or we can change the order of in `.datasets_to_data`

```r
  code <- c(
    get_rcode_str_install(),
    get_rcode_libraries(),
    get_datasets_code(datanames, datasets, hashes)
  )


  data <- do.call(
    teal.data::teal_data,
    args = c(data, code = list(code), join_keys = list(datasets$get_join_keys()[datanames]))
  )
````

So that `teal.data::teal_data()` is called and put into `get_datasets_code()`. Happy to discuss

chlebowa commented on R/utils.R:
```diff
@@ -68,7 +68,7 @@ teal_data_to_filtered_data <- function(x, datanames = teal_data_datanames(x)) {
     join_keys = teal.data::join_keys(x)
   )
   # Piggy-back entire pre-processing code so that filtering code can be appended later.
```
```suggestion
  # Piggy-back pre-processing code for datasets of interest so that filtering code can be appended later.
```

chlebowa reviewed (changes requested):
Needs work after all.

May involve rewriting the `module_datasets` function or a completely different approach.

m7pr commented:
Coolio. Just pasting the code we use to test the solution

```r
app <- init(
  data = teal_data(
    IRIS = iris, 
    MTCARS = mtcars, 
    code = "IRIS <- iris; x <-5; MTCARS <- mtcars"
  ),
  modules = modules(
    example_module(label = "MTCARS only", datanames = "MTCARS"),
    example_module(label = "IRIS only", datanames = "IRIS")
  ),
  filter = teal_slices(
    teal_slice("MTCARS", "mpg"),
    teal_slice("IRIS", "Species"),
    module_specific = TRUE,
    mapping = list(
      global_filters = c("MTCARS mpg", "IRIS Species")
    )
  )
)
if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

chlebowa commented:
It took me a while but:
I added a new file, `modules_datasets.R`. You can move the contents as you wish. The file contains
1. `modules_structure` - small functions that returns a nested list that reflects the structure of a `teal_modules`
2. `modules_datasets` - rather large function that returns list of `FilteredData` objects that reflects the shape of `teal_modules`
3. some testing, which can be easily adapted to the proper place in `module_teal.R` (uncomment)

chlebowa commented on R/modules_datasets.R:
```diff
@@ -0,0 +1,106 @@
+    slices <- shiny::isolate({
+      Filter(function(x) x$id %in% attr(filter, "mapping")$global_filters, filter)
+    })
+    singleton$set_filter_state(slices)
```
Yes, sorry about that. Fixed.

m7pr commented:
# Testing in the script

Hey, for `module_specific = TRUE` in your testing scenario I do get:

```r
$two
$two$example_two
[1] "mtcars cyl"

$two$three
$two$three$example_three
list()

$two$three$example_four
[1] "mtcars mpg" "mtcars cyl"



$example_one
[1] "iris Species"
```

Which looks proper. Of course for `module_specific = FALSE` I get

```r

$two
$two$example_two
[1] "mtcars cyl"

$two$three
$two$three$example_three
[1] "mtcars cyl"

$two$three$example_four
[1] "mtcars cyl"



$example_one
[1] "mtcars cyl"
```

which is expected. This part is great.


# Testing in the package

I tried to substitute `module_datasets()` with `modules_datasets()` https://github.com/insightsengineering/teal/pull/1115/files but got this error

```r
Warning in private$set_choices(slice$choices) :
  Choices adjusted (some values outside of variable range). Varname: mpg, dataname: MTCARS.
Warning: Error in ui_nested_tabs.teal_module: Assertion on 'datasets' failed: Must inherit from class 'FilteredData', but has class 'NULL'.
  3: runApp
  2: print.shiny.appobj
  1: <Anonymous>
```

while using 

```r
app <- init(
  data = teal_data(
    IRIS = iris, 
    MTCARS = mtcars, 
    code = "IRIS <- iris; x <-5; MTCARS <- mtcars"
  ),
  modules = modules(
    example_module(label = "all datasets", datanames = "all"),
    example_module(label = "MTCARS only", datanames = "MTCARS"),
    example_module(label = "IRIS only", datanames = "IRIS")
  ),
  filter = teal_slices(
    teal_slice("MTCARS", "mpg"),
    teal_slice("IRIS", "Species"),
    module_specific = TRUE,
    mapping = list(
      global_filters = c("MTCARS mpg", "IRIS Species")
    )
  )
)
if (interactive()) {
  shinyApp(app$ui, app$server)
}
```

m7pr commented:
Hey @chlebowa I think we are good to merge.
We just need to 
a) remove commented code from `modules_datasets()`
b) decide where this functions lives
c) add tests for it?

chlebowa commented:
Or do you want me to do it? :wink:

m7pr commented:
@chlebowa sorry for the late reply. I can handle! No worries :)

m7pr commented:
Hey @chlebowa I cleaned up your implementation and moved the function to utils.R. Added small roxygen documentation. Used your testing scenario in a test in testtthat folder. Lastly change the name of `filter` parameter, to `filters` so it plays with other parameters called `filters` in other functions that take `teal_slices` as an input.

chlebowa commented:
> Lastly change the name of `filter` parameter, to `filters` so it plays with other parameters called `filters` in other functions that take `teal_slices` as an input.

Excellent :ok_hand:

chlebowa commented on R/utils.R:
```diff
@@ -220,6 +220,87 @@ check_filter_datanames <- function(filters, datanames) {
 }
 
+
+#' Extract datasets for modules
```
```suggestion
#' Create filterable data for modules
#' 
#' Converts input data to a `FilteredData` object(s) to allow filtering before passing data to individual modules.
```

chlebowa commented on R/utils.R:
```diff
@@ -220,6 +220,87 @@ check_filter_datanames <- function(filters, datanames) {
+      Filter(x = filters, f = function(x) {
+        x$dataname %in% datanames &&
+          (x$id %in% attr(filters, "mapping")$global_filters ||
+            x$id %in% unique(unlist(attr(filters, "mapping")[modules$label]))) # nolint
```
Please specify linter exception.

m7pr replied:
this is this ridiculous styler vs lintr infinite war : D

chlebowa replied:
I don't want an explanation, I want a specification. Which linter was turned off for this line?

m7pr replied:
got it. was trying to turn off `indentation_linter` https://github.com/insightsengineering/teal/pull/1118
put a PR for this

chlebowa replied:
pertect

chlebowa commented on R/utils.R:
```diff
@@ -220,6 +220,87 @@ check_filter_datanames <- function(filters, datanames) {
+  checkmate::assert_r6(filtered_data_singleton, "FilteredData")
+
+  if (!isTRUE(attr(filters, "module_specific"))) {
+    slices <- shiny::isolate({
```
```suggestion
    # subset global filters
    slices <- shiny::isolate({
```

chlebowa commented on tests/testthat/test-utils.R:
```diff
@@ -64,6 +64,63 @@ test_that("teal_data_datanames returns datanames which are set by teal.data::dat
+    `example one` = "iris Species"
+  )
+
+  testthat::expect_equal(modules_structure, expected_structure)
```
`expect_identical`?

chlebowa reviewed (approved):
Some minor comments, otherwise ready.
